### PR TITLE
Tools to run all tests automatically.

### PR DESCRIPTION
**Pull request by pissang**
### Problem to Solve

Currently our testing is heavily depends on the nearly 400 test files in test folder. They are all HTML test files and we need to run them one by one manually when doing regression test, which is very inefficient. 

So this pull request is about to solve this problem by providing a tool to run all tests automatically and compared the screenshots to the latest stable version to check if anything bad happens. 

NOTE: It IS NOT designed for unit testing. It's only for visual regression testing.

### Current Progress

A simple screenshot.

MacHi 2019-09-02 23-48-23

It includes:

+ A CLI to run target tests with [puppeteer](https://github.com/GoogleChrome/puppeteer) and record all screenshots, logs, errors. It will compare the screenshots between the current developing version and the latest stable version.
+ A dashboard can control the running of tests and generate reports.
+ A server which connect dashboard and CLI. It will receive the commands from dashboard and invoke CLI to run given tests. The result from CLI will be sent back to the dashboard.

### How to Use

1. Install necessary dependencies

```bash
npm install
```

2. Start the server

```bash
npm run test:visual
```

3. Open the dashboard in the browser http://localhost:8866/test/runTest/client/index.html

### TODO & Further Discussion

There are still many things needs to be done.

1. The most important part is interaction testing. Currently it only compares the final screenshot. But many tests are created for interaction testing, which means we need to simulate the user behavior in the puppeteer. There are two solutions about this:

    1. Write codes to control interaction manually, like click, drag, dispatchAction, etc. It's more precise because we know when to generate a screenshot.
    2. Write a recorder to record the user interaction. Playback when running tests. We don't need to write code, but perhaps we can't control when to screenshot so precisely.

Both of them needs a lot of work for 400 tests. 

2. Secondly, I can't find a good way to mock time. It makes testing animations hard because builtin timer are not so accurate. There is always 1-2ms difference between two run, or even worse. The lucky thing is I found a library can do this thing https://github.com/sinonjs/lolex/ . But I was not lucky enough because it will cause some tests hang, and the animation tests are still wrong.

3. Currently it only tests canvas renderer, we can add svg renderer testing on all the tests files.

If you have any questions or suggestions. Please leave your comments. Thanks.

### CODE REVIEW

**Comment by pissang** [comment]
I removed puppeteer from devDependencies. It will download a binary executable file, which may cause install failed in China.

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -77,7 +77,15 @@ Then, open the test cases under `~/workspace/echarts/test` in Web browser. You c
...
 ![Chrome inspect](../asset/contributing-inspect.png)
 
-### 4. Make a pull request
+### 4. Run test
+
+```bash
+node test/runTest/cli.js
+```
+
+It will run all the test cases under `~/workspace/echarts/test` automatically to compare with the previous version. You can use this to check if your code bring some breaking change.
```
Comment (Ovilia) [inline .github/CONTRIBUTING.md]: `~/workspace`?

File: test/runTest/blacklist.js [hunk 2]
```diff
@@ -0,0 +1,4 @@
+module.exports = [
+    '-cases.html',
+    'geo-random-stream.html'
+];
```
Comment (Ovilia) [inline test/runTest/blacklist.js]: A blank line at the end of every file?

File: test/runTest/runtime/main.js [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+//         return new NativeDate(params);
+//     }
+// }
+// MockDate.prototype = new Date();
+// // Fixed date
+// window.Date = MockDate;
+
+export function createScreenshotTest (desc, elementQuery, waitTime) {
+
+}
```
Comment (Ovilia) [inline test/runTest/runtime/main.js]: `;`?

**Comment by pissang** [comment]
Hi @Ovilia , I fixed the coding style you mentioned

**Pull request merged**